### PR TITLE
Proposed fix for #1579: adjusting stop_request and quit_request

### PR DESCRIPTION
**Pull request by drraid**
Slightly changed how stop_request and quit_request are set and 
checked. First, this adds uc_emu_soft_stop(), which now has all
the same logic that was previously in uc_emu_stop(), except it
does not set uc->stop_request.  The uc_emu_stop() function now
calls uc_emu_soft_stop() and then sets uc->stop_request. The
point of making uc_emu_soft_stop its own function was to split
out the logic of using stop_request vs quit_request: places
that just "quit" (as in flush but then continue emulating) still
set the uc->quit_request flag, but now call uc_emu_soft_stop()
instead of uc_emu_stop(). The logic in softmmu/cpus.c now only
checks if uc->stop_request is set to determine if it should
break out of the loop.

The target architectures have been updated to call the new
uc_emu_soft_stop instead of uc_emu_stop for their "quits".

TL;DR stop_request and quit_request are now one-or-the-other
instead of one-and-maybe-the-other.

This is to address #1579 

NOTE: This is my first time looking at the inner workings of
Unicorn, and a grown-up should review this carefully because
I probably did this wrong ;). I wanted to attempt a fix instead 
of simply filing a bug with a repro. I will not be offended if my
fix is partially or completely rejected!

### CODE REVIEW

**Comment by wtdcode** [comment]
Both #1579 and #1173 have been known bugs for Unicorn for quite a long time (since I get familiar with this project #1356).

However, I do think it's not a critical bug or design flaw since users still have easy workaround to achieve what they would like.

Could you come up with a scenario where this API is really essential? Note we are extremely cautious about adding/removing an existing API.

**Comment by drraid** [comment]
I'm going to close this pull request because I'm pretty sure I broke other stuff, and I will look at it more to propose a better fix that doesn't break stuff. 

I hadn't seen #1356 - my apologies for yet another duplicate.

I don't want to add or remove an API, but I think it would be good if the API behaved as expected. In my scenario I have created a library to wrap Unicorn and provide functionality like breakpoints, which is how I ran into this issue. When implementing a breakpoint I want to notify the user that the breakpoint has been reached without executing the instruction, and when notifying the user of the breakpoint I also want to have the PC point correctly to the instruction where the breakpoint is. I have implemented the breakpoint the same way I have in debuggers: catch the INT3, write back the original byte, set the PC back to that byte, then notify the user. If a user wants to stop execution (or emulation) at this point, they should have the choice to do so. 

I could hack around it by fully stopping the emulator when I receive the INT3, and then fully starting again if the user indicates that the breakpoint should proceed, but then that only solves the expected behavior for the breakpoint case, and doesn't change the expectation of uc_emu_stop(). 

I totally understand the desire to not change APIs, and especially understand your caution in accepting pull requests as I'm pretty sure my PR is broken :), but a correct fix feels like it would be aligned with user expectations of the API.
